client/signup/controller.js

```diff
@@ -287,6 +287,11 @@ export default {
 			userLoggedIn
 		);
 
+		// The woocommerce-install flow needs to support site switching.
+		if ( 'woocommerce-install' === flowName ) {
+			initialContext = context;
+		}
```

Comment:
@Automattic/explorers This is rather hacky but might fix our site switching problem?

Before we commit to this we should probably double-check with someone who knows signup whether this will explode or not.

Reply:
It seems to fix our issue! 👍

Reply:
> The woocommerce-install flow needs to support site switching.

I think the site switching problem is a general problem in the signup framework, not just for `woocommerce-install`. For instance the launch flow: the user could begin the launch flow, back all the way out, switch to another site and launch it.

Is the underlying issue that the `?site` query param isn't being propagated as you traverse through the steps? I'm guessing that since the `?site` param disappears, the signup framework ends up falling back to the old param that's still saved in `initialContext`. We had a similar issue in the `setup-site` flow that @roo2 addressed by retaining the query param as the user moves between steps (https://github.com/Automattic/wp-calypso/pull/58498). I think you would have inherited this behaviour too except you're using a `site` param instead of `siteSlug`.

Reply:
Thank you, @p-jackson for the 👀!

> Is the underlying issue that the ?site query param isn't being propagated as you traverse through the steps?

Perhaps, addressing that could work for us as well. It's probably worth a try, but I wonder if the current solution we have drafted would be adequate? Any thoughts on our current approach?

Reply:
We could swap to siteSlug easy enough, it would just be visible through every step in the flow.

Although siteSlug still has a precedence problem in that it prefers local storage over what's in the url. Adding the same propogation for `site=` might be the way forward?

Reply:
> Is the underlying issue that the ?site query param isn't being propagated as you traverse through the steps? 

The issue seems to be that the changed ?site param never gets picked up from `context` because it only looks at `initialContext`. So even if we maintained the parameter, it would still not be used.

Reply:
^ take with grain of salt, not sure I know enough about this to be sure.

Reply:
There are plenty of other flow specific hacks in the signup code (i.e. places where we do custom things based on the value of `flowName`), so I can't say that this is hackier than anything else we do in signup. I can't think of anything in the `woocommerce-install` flow where this approach is going to introduce a bug, but there are lots of different modes/hacks/features of the signup framework, and it's the unknown unknowns that I'm always paranoid about.

I'm not very familiar with `initialContext` and I don't know what other experiences might have come to rely on it. It looks to have been added in #18557 to facilitate social signup.

In saying that, if you're confident in your manual testing&mdash;checking all the other ways that it's possible to navigate into, out of and around the `woocommerce-install` flow&mdash;then I think it's adequate. It won't affect any of the existing flows.

> // The woocommerce-install flow needs to support site switching.

I think this comment is misleading though. It could lead readers to believe that they should also add their flow to this if statement when their flow needs to support site switching. I don't think it's an adequate solution for all flows in general.

Reply:
> > // The woocommerce-install flow needs to support site switching.
> I think this comment is misleading though. It could lead readers to believe that they should also add their flow to this if statement 

I had not considered that, good catch

Reply:
> In saying that, if you're confident in your manual testing—checking all the other ways that it's possible to navigate into, out of and around the woocommerce-install flow—then I think it's adequate. It won't affect any of the existing flows.

The optional checkout step that returns back into the flow would be the one to check.